File: tools/gemini/fast_ask.py
"""
Gemini Fast Ask - Speed-optimized text chat using a persistent browser daemon.

Keeps Chrome alive between calls. Reuses the same conversation for up to 20
prompts before starting fresh. Uses a dedicated "ask" tab in the daemon so
it never interferes with ai-gem or other tools.

First call:  ~10-15s (launches browser + navigates)
Subsequent:  ~4-5s (connect + send prompt + wait response)

Usage:
    poetry run python -m tools.gemini.fast_ask "What is Python?"
    poetry run python -m tools.gemini.fast_ask --quiet "2+2"
    poetry run python -m tools.gemini.fast_ask --new "start fresh conversation"
"""

import argparse
import functools
import sys
import time

from tools.gemini.base import GeminiBase
from tools.gemini.browser import BrowserDaemon, GEMINI_URL

_print = functools.partial(print, flush=True)

MAX_PROMPTS_PER_CONVERSATION = 20
ROLE = "ask"


class GeminiFastAsk(GeminiBase):
    """Speed-optimized Gemini text chat using persistent browser daemon."""

    TOOL_NAME = None
    TOOL_LABEL = "fast_ask"
    DEFAULT_EXT = ".txt"

    def __init__(self, quiet=False):
        super().__init__(quiet=quiet)
        self._daemon = BrowserDaemon(quiet=quiet)

    def run_fast(
        self,
        prompt: str,
        force_new: bool = False,
        file_path: str | None = None,
    ):
        """
        Send a prompt and get a response using the persistent browser.

        Parameters
        ----------
        prompt : str
            The text prompt to send.
        force_new : bool
            Force a new conversation even if under the prompt limit.
        file_path : str, optional
            Path to a text file whose content will be used as the prompt.
        """
        # If file_path provided, read content
        if file_path:
            try:
                with open(file_path, encoding="utf-8") as f:
                    file_content = f.read().strip()
                if prompt:
                    prompt = f"{prompt}\n\n{file_content}"
                else:
                    prompt = file_content
            except Exception as e:
                self.log(f"  ERROR: Could not read file: {e}")
                return None

        try:
            # Step 1: Ensure browser daemon is running
            state = self._daemon.ensure_browser()
            count = state.get("ask_count", 0)

            # Step 2: Decide whether to reuse or start new conversation
            need_new = force_new or count >= MAX_PROMPTS_PER_CONVERSATION

            if need_new:
                self.log("  Starting new conversation...")
                browser, page = self._daemon.new_conversation(ROLE, GEMINI_URL)
                count = 0
            else:
                browser, page = self._daemon.get_page(ROLE, GEMINI_URL)

            # Step 3: Check login (skip on 2nd+ calls - already verified)
            if count == 0:
                if not self.is_logged_in(page):
                    self.log("  ERROR: Not logged in. Run: ai-ask --login \"hello\"")
                    self._daemon.release()
                    return None
                # Dismiss dialogs only on first use
                self._dismiss_dialogs(page)

            # Step 4: Send prompt
            self.log(f"  Sending prompt ({count + 1}/{MAX_PROMPTS_PER_CONVERSATION})...")
            success = self._send_prompt_fast(page, prompt)
            if not success:
                self.log("  ERROR: Failed to send prompt")
                self._daemon.release()
                return None

            # Step 5: Wait for response
            self.log("  Waiting for response", end="")
            response = self._wait_for_text_response(page)

            # Step 6: Update state
            BrowserDaemon.update_state(ask_count=count + 1)

            # Print response
            if response:
                self.log("")  # newline after dots
                _print(response)
            else:
                self.log("\n  (No response captured)")

            self._daemon.release()
            return response

        except Exception as e:
            self.log(f"  ERROR (daemon): {e}")
            try:
                self._daemon.release()
            except Exception:
                pass
            # If daemon failed, fall back to standard mode
            self.log("  Falling back to standard mode (no daemon)...")
            try:
                BrowserDaemon.stop()
            except Exception:
                pass
            return self._fallback_run(prompt)

    def _fallback_run(self, prompt):
        """Fall back to standard GeminiAsk when daemon is unavailable."""
        from tools.gemini.ask import GeminiAsk

        tool = GeminiAsk(quiet=self._quiet)
        return tool.run(prompt=prompt, headless=True)

    def _send_prompt_fast(self, page, prompt_text):
        """Fast prompt sending - minimal sleeps for speed."""
        try:
            input_el = page.locator('.ql-editor[contenteditable="true"]').first
            input_el.click(timeout=5000)

            if len(prompt_text) > 200:
                input_el.fill(prompt_text)
            else:
                page.keyboard.type(prompt_text)
            time.sleep(0.3)

            # Click send button (avoid matching the "Stop response" button)
            send_btn = page.locator(
                'button[aria-label="Send message"]:not([aria-label*="Stop"])'
            ).first
            try:
                if send_btn.is_visible(timeout=3000):
                    send_btn.click(timeout=5000)
                    time.sleep(0.3)
                    return True
            except Exception:
                pass

            # Fallback: Enter key (most reliable)
            page.keyboard.press("Enter")
            time.sleep(0.3)
            return True

        except Exception as e:
            self.log(f"  ERROR sending: {e}")
            return False

    def _wait_for_text_response(self, page, timeout_seconds=120):
        """Wait for text response - fast polling version."""
        start = time.time()
        last_text = ""
        stable_count = 0

        while time.time() - start < timeout_seconds:
            try:
                result = page.evaluate("""() => {
                    const stopBtn = document.querySelector('button[aria-label*="Stop"]');
                    const isGenerating = stopBtn && stopBtn.offsetParent !== null;

                    const modelResp = document.querySelectorAll('model-response');
                    let text = '';
                    if (modelResp.length > 0) {
                        text = (modelResp[modelResp.length - 1].innerText || '').trim();
                    }
                    if (!text) {
                        const msgs = document.querySelectorAll('message-content');
                        if (msgs.length > 0) {
                            text = (msgs[msgs.length - 1].innerText || '').trim();
                        }
                    }
                    return { g: isGenerating, t: text };
                }""")

                text = result.get("t", "")
                generating = result.get("g", False)

                if text and len(text) > 5 and not generating:
                    if text == last_text:
                        stable_count += 1
                        if stable_count >= 2:
                            return self._clean_response(text)
                    else:
                        stable_count = 0
                    last_text = text

            except Exception:
                pass

            self.log(".", end="")
            time.sleep(1)

        if last_text:
            return self._clean_response(last_text)
        return None

    # Keep compatibility with base class
    def _wait_for_response(self, page, **kwargs):
        return self._wait_for_text_response(page)


def main():
    parser = argparse.ArgumentParser(
        description="Gemini Fast Ask - Speed-optimized text chat",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  poetry run python -m tools.gemini.fast_ask "What is Python?"
  poetry run python -m tools.gemini.fast_ask --quiet "2+2"
  poetry run python -m tools.gemini.fast_ask --new "fresh conversation"
  poetry run python -m tools.gemini.fast_ask --stop   # Kill browser daemon
  poetry run python -m tools.gemini.fast_ask --status # Check daemon status

For Gem usage, use: ai-gem (tools.gemini.fast_gem)
""",
    )
    parser.add_argument("prompt", nargs="?", default=None, help="The prompt to send")
    parser.add_argument("--file", type=str, default=None, help="Read prompt from text file")
    parser.add_argument("--new", action="store_true", default=False, help="Force new conversation")
    parser.add_argument("--quiet", action="store_true", default=False, help="Only print response")
    parser.add_argument("--stop", action="store_true", default=False, help="Stop browser daemon")
    parser.add_argument("--status", action="store_true", default=False, help="Show daemon status")
    parser.add_argument("--login", action="store_true", default=False, help="Force re-login (uses slow mode)")

    args = parser.parse_args()

    if args.stop:
        BrowserDaemon.stop()
        return

    if args.status:
        BrowserDaemon.status()
        return

    if args.login:
        # Stop daemon first (it locks the profile)
        BrowserDaemon.stop()
        # Fall back to regular GeminiAsk for login flow
        from tools.gemini.ask import GeminiAsk
        tool = GeminiAsk(quiet=args.quiet)
        response = tool.run(
            prompt=args.prompt or "hello",
            headless=False,
            force_login=True,
        )
        if not response:
            sys.exit(1)
        return

    if not args.prompt and not args.file:
        parser.print_help()
        sys.exit(1)

    tool = GeminiFastAsk(quiet=args.quiet)
    response = tool.run_fast(
        prompt=args.prompt or "",
        force_new=args.new,
        file_path=args.file,
    )

    if not response:
        sys.exit(1)


if __name__ == "__main__":
    main()

File: go-dev-tools/internal/gemini/fast_ask.go
package gemini

import (
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/go-rod/rod"
	"github.com/prd-nguyen-trancong/gemini-tools/internal/browser"
)

const (
	MaxPromptsPerConversation = 20
	RoleAsk                   = "ask"
)

// FastAsk provides speed-optimized Gemini text chat using the persistent daemon.
type FastAsk struct {
	Base
	daemon *browser.Daemon
}

// NewFastAsk creates a new FastAsk instance.
func NewFastAsk(quiet bool) *FastAsk {
	return &FastAsk{
		Base:   Base{Quiet: quiet},
		daemon: browser.NewDaemon(quiet),
	}
}

// RunFast sends a prompt and returns the response using the persistent browser.
func (fa *FastAsk) RunFast(prompt string, forceNew bool, filePath string) (string, error) {
	// If filePath provided, read content
	if filePath != "" {
		data, err := os.ReadFile(filePath)
		if err != nil {
			return "", fmt.Errorf("read file: %w", err)
		}
		fileContent := strings.TrimSpace(string(data))
		if prompt != "" {
			prompt = prompt + "\n\n" + fileContent
		} else {
			prompt = fileContent
		}
	}

	// Step 1: Ensure browser daemon is running
	state, err := fa.daemon.EnsureBrowser()
	if err != nil {
		return "", fmt.Errorf("ensure browser: %w", err)
	}
	count := state.AskCount

	// Step 2: Decide whether to reuse or start new conversation
	needNew := forceNew || count >= MaxPromptsPerConversation

	if needNew {
		fa.LogLn("  Starting new conversation...")
		_, page, err := fa.daemon.NewConversation(RoleAsk, browser.GeminiURL)
		if err != nil {
			return "", fmt.Errorf("new conversation: %w", err)
		}
		count = 0

		// Check login on fresh conversations
		if !fa.IsLoggedIn(page) {
			fa.LogLn("  ERROR: Not logged in. Run: go-ai-ask --login \"hello\"")
			fa.daemon.Release()
			return "", fmt.Errorf("not logged in")
		}
		fa.DismissDialogs(page)

		return fa.sendAndWait(page, prompt, count)
	}

	// Reuse existing conversation
	_, page, err := fa.daemon.GetPage(RoleAsk, browser.GeminiURL)
	if err != nil {
		return "", fmt.Errorf("get page: %w", err)
	}

	// Check login on first call
	if count == 0 {
		if !fa.IsLoggedIn(page) {
			fa.LogLn("  ERROR: Not logged in. Run: go-ai-ask --login \"hello\"")
			fa.daemon.Release()
			return "", fmt.Errorf("not logged in")
		}
		fa.DismissDialogs(page)
	}

	return fa.sendAndWait(page, prompt, count)
}

func (fa *FastAsk) sendAndWait(page *rod.Page, prompt string, count int) (string, error) {
	// Count existing responses BEFORE sending (so we detect the NEW one)
	prevCount := countModelResponses(page)

	// Send prompt
	fa.LogLn("  Sending prompt (%d/%d)...", count+1, MaxPromptsPerConversation)
	if !fa.SendPromptFast(page, prompt) {
		fa.daemon.Release()
		return "", fmt.Errorf("failed to send prompt")
	}

	// Wait for response (pass prevCount so we only look at NEW responses)
	fa.Log("  Waiting for response")
	response := fa.WaitForTextResponse(page, 120, prevCount)

	// Update state
	browser.UpdateState(map[string]interface{}{"ask_count": count + 1})

	fa.daemon.Release()

	if response != "" {
		fa.LogLn("")
		fmt.Println(response)
	} else {
		fa.LogLn("\n  (No response captured)")
	}
	return response, nil
}

// Stop kills the browser daemon.
func (fa *FastAsk) Stop() {
	browser.Stop()
}

// Status prints daemon status.
func (fa *FastAsk) Status() {
	browser.Status()
}

// Login performs the login flow (opens VISIBLE browser for Google login).
//
// Flow:
//  1. Stop existing daemon → launch visible browser (on-screen) for user login
//  2. User logs in (passkey may crash browser - that's OK, cookies persist)
//  3. Stop visible browser → restart as headless daemon
//  4. Verify login in headless mode and send initial prompt
func (fa *FastAsk) Login(prompt string) (string, error) {
	browser.Stop()
	time.Sleep(2 * time.Second)

	fa.LogLn("============================================================")
	fa.LogLn("  Go Gemini Login")
	fa.LogLn("============================================================")
	fa.LogLn("  Profile: .browser-data/gemini-go/")
	fa.LogLn("  NOTE: Passkey login may close the browser - that's OK!")
	fa.LogLn("============================================================")

	// ── Step 1: Launch VISIBLE browser for manual login ──
	fa.LogLn("\n[Step 1] Opening visible browser for login...")
	loginResult := fa.waitForManualLogin()

	if loginResult == "BROWSER_CLOSED" {
		fa.LogLn("  Browser closed (passkey flow). Cookies should be saved.")
	} else if loginResult == "LOGGED_IN" {
		fa.LogLn("  Login detected in visible browser!")
	} else {
		return "", fmt.Errorf("login failed or timed out")
	}

	// ── Step 2: Stop visible browser gracefully and restart as headless daemon ──
	fa.LogLn("\n[Step 2] Flushing cookies and stopping visible browser...")
	// Give Chrome time to flush cookies/session to disk before killing
	time.Sleep(5 * time.Second)
	browser.StopGraceful()
	time.Sleep(3 * time.Second)

	headlessDaemon := browser.NewDaemon(false)
	_, err := headlessDaemon.EnsureBrowser()
	if err != nil {
		return "", fmt.Errorf("start headless daemon: %w", err)
	}

	_, page, err := headlessDaemon.GetPage(RoleAsk, browser.GeminiURL)
	if err != nil {
		return "", fmt.Errorf("get page: %w", err)
	}
	time.Sleep(3 * time.Second)

	if !fa.IsLoggedIn(page) {
		fa.LogLn("  Login not detected in headless mode. Please try again.")
		headlessDaemon.Release()
		return "", fmt.Errorf("login not persisted to headless mode")
	}

	fa.LogLn("  Login verified in headless mode!")
	fa.DismissDialogs(page)

	// ── Step 3: Send initial prompt ──
	return fa.loginSendPrompt(headlessDaemon, page, prompt)
}

// waitForManualLogin launches a visible browser and waits for login or browser crash.
// Returns: "LOGGED_IN", "BROWSER_CLOSED", or "" (timeout).
func (fa *FastAsk) waitForManualLogin() string {
	daemon := browser.NewDaemon(false)
	_, err := daemon.EnsureBrowserVisible()
	if err != nil {
		fa.LogLn("  ERROR: Could not launch visible browser: %v", err)
		return ""
	}

	// Connect and get the first page (already navigated to Gemini by EnsureBrowserVisible)
	brow, err := daemon.Connect()
	if err != nil {
		fa.LogLn("  ERROR: Could not connect to browser: %v", err)
		return ""
	}

	// Wait for at least one page to be available (Chrome may take a moment)
	var page *rod.Page
	for attempt := 0; attempt < 15; attempt++ {
		pages, pErr := brow.Pages()
		if pErr == nil && len(pages) > 0 {
			page = pages[0]
			break
		}
		time.Sleep(1 * time.Second)
	}
	if page == nil {
		fa.LogLn("  ERROR: No pages found in browser after waiting")
		return ""
	}

	fa.LogLn("  Please log in to your Google account in the browser window.")
	fa.LogLn("  (Passkey login may close the browser - that's OK!)")

	for i := 0; i < 300; i++ {
		// Check if Chrome process is still alive (fastest check)
		state := browser.LoadState()
		if state.PID > 0 && !browser.IsPIDAlive(state.PID) {
			fa.LogLn("  Chrome process exited (passkey flow detected).")
			return "BROWSER_CLOSED"
		}

		// Check if CDP port is still open
		if !browser.IsPortOpen(browser.CDPPort) {
			fa.LogLn("  CDP port closed (passkey flow detected).")
			return "BROWSER_CLOSED"
		}

		// Check if browser/page is still alive
		_, infoErr := page.Info()
		if infoErr != nil {
			errMsg := strings.ToLower(infoErr.Error())
			if strings.Contains(errMsg, "closed") ||
				strings.Contains(errMsg, "crash") ||
				strings.Contains(errMsg, "connection") ||
				strings.Contains(errMsg, "websocket") ||
				strings.Contains(errMsg, "eof") ||
				strings.Contains(errMsg, "target") {
				fa.LogLn("  Browser closed (passkey flow detected).")
				return "BROWSER_CLOSED"
			}
		}

		// Check if logged in (try all pages in case of redirect)
		allPages, pErr := brow.Pages()
		if pErr == nil {
			for _, pg := range allPages {
				if fa.IsLoggedIn(pg) {
					daemon.Release()
					return "LOGGED_IN"
				}
			}
		}

		if i > 0 && i%15 == 0 {
			fa.LogLn("  Still waiting for login... (%ds)", i)
		}
		time.Sleep(1 * time.Second)
	}

	daemon.Release()
	fa.LogLn("  Timeout waiting for login.")
	return ""
}

// loginSendPrompt sends the initial prompt after login and returns the response.
func (fa *FastAsk) loginSendPrompt(daemon *browser.Daemon, page *rod.Page, prompt string) (string, error) {
	fa.LogLn("\n[Final] Sending prompt: %s", prompt)

	if !fa.SendPromptFast(page, prompt) {
		fa.LogLn("  Warning: Could not send prompt, but login was saved.")
		daemon.Release()
		fa.LogLn("\n  Login session saved to .browser-data/gemini-go/")
		fa.LogLn("  You can now use go-ai-ask normally.")
		return "", nil
	}

	fa.Log("  Waiting for response")
	response := fa.WaitForTextResponse(page, 120, 0)

	browser.UpdateState(map[string]interface{}{"ask_count": 1})
	daemon.Release()

	fa.LogLn("\n============================================================")
	fa.LogLn("  Login successful! Session saved to .browser-data/gemini-go/")
	fa.LogLn("  You can now use go-ai-ask normally (headless mode).")
	fa.LogLn("============================================================")

	if response != "" {
		fa.LogLn("")
		fmt.Println(response)
	}
	return response, nil
}
